thunderfish/eventdetection.py

```diff
@@ -145,6 +145,121 @@ def detect_peaks(data, threshold):
     return np.asarray(peaks_list, dtype=np.int), np.asarray(troughs_list, dtype=np.int)
 
 
+@jit(nopython=True)
+def detect_peaks_c(data, threshold):
+    """
+    Detect peaks and troughs using a fixed, relative threshold according to
+    Bryan S. Todd and David C. Andrews (1999): The identification of peaks in physiological signals.
+    Computers and Biomedical Research 32, 322-335.
+
+    Parameters
+    ----------
+    data: array
+        An 1-D array of input data where peaks are detected.
+    threshold: float or array
+        A positive number or array of numbers setting the detection threshold,
+        i.e. the minimum distance between peaks and troughs.
+        In case of an array make sure that the threshold does not change faster
+        than the expected intervals between peaks and troughs.
+    
+    Returns
+    -------
+    peak_array: array of ints
+        A list of indices of detected peaks.
+    trough_array: array of ints
+        A list of indices of detected troughs.
+
+    Raises
+    ------
+    ValueError: If `threshold <= 0`.
+    IndexError: If `data` and `threshold` arrays differ in length.
+    """
+
+    thresh_array = True
+    thresh = 0.0
+    #if np.isscalar(threshold):
+    thresh_array = False
+    thresh = threshold
+    if threshold <= 0:
+        raise ValueError('input argument threshold must be positive!')
+    #elif len(data) != len(threshold):
+    #    raise IndexError('input arrays data and threshold must have same length!')
+
+
+    # initialize:
+    direction = 0
+    min_inx = 0
+    max_inx = 0
+    min_value = data[0]
+    max_value = min_value
+
+    peaks_list = np.zeros(len(data))
```

Comment:
Should be
```
peaks_list = np.zeros(len(data)//2, dtype=int)
```
These are used as indices and indices must be ints and not floats (the default of np.zeros()).

Same for `troughs_list`

Reply:
If I do this, I have to define dtype=numba.int64, as numba has more specific types than python. Then when using this:
try:
    from numba import jit
except ImportError:
    def jit(nopython):
        def decorator_jit(func):
            return func
        return decorator_jit
it wouldn't work when numba is not imported. I tried using try - except again when defining np.zeros dtype, but as with numba, compilation is done beforehand, it has to understand both statements upfront. As it does not understand the python int, and also function such as np.isscalar also do not work in numba, the numba functions are rather limited. I found a workaround using two functions (check last commit.)